### PR TITLE
New logic to allow altering of SpySat status via [Upgrade] 

### PR DESCRIPTION
**Pull request by DOOMGUY2K**
Upgrades in Vanilla can not affect

[SpySat]

This fixes that. 

[Rules]md.ini

[Building] 
PowersUp.Buildings=NARADR
PowersUpBuilding=NARADR
PowersUpToLevel=-1
SpySat=yes

[ParentBuilding]
Upgrades=1

### CODE REVIEW

File: src/Misc/Hooks.BugFixes.cpp [hunk 1]
```diff
@@ -636,6 +636,25 @@ DEFINE_HOOK(0x51A996, InfantryClass_PerCellProcess_KillOnImpassable, 0x5)
...
+// Fixes broken Upgrade logic to allow SpySat=Yes Code=Otamma / Tested DOOM3DGUY
+DEFINE_HOOK(0x508F82, HouseClass_AI_checkSpySat_IncludeUpgrades , 0x6) {
+
+    enum { AdvanceLoop = 0x508FF6, Continue = 0x508F91 };
+
+    GET(BuildingClass const*, pBuilding, ECX);
+
+    if (!pBuilding->Type->SpySat) {
+        for (const auto& pUpGrade : pBuilding->Upgrades) {
+            if (pUpGrade && pUpGrade->SpySat)
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:648]:
```suggestion
    if (!pBuilding->Type->SpySat)
    {
        for (const auto& pUpgrade : pBuilding->Upgrades)
        {
            if (pUpGrade && pUpgrade->SpySat)
```

File: src/Misc/Hooks.BugFixes.cpp [hunk 2]
```diff
@@ -636,6 +636,25 @@ DEFINE_HOOK(0x51A996, InfantryClass_PerCellProcess_KillOnImpassable, 0x5)
 	return SkipKilling;
 }
 
+// Fixes broken Upgrade logic to allow SpySat=Yes Code=Otamma / Tested DOOM3DGUY
+DEFINE_HOOK(0x508F82, HouseClass_AI_checkSpySat_IncludeUpgrades , 0x6) {
+
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:641]:
```suggestion
DEFINE_HOOK(0x508F82, HouseClass_AI_CheckSpySat_IncludeUpgrades, 0x6)
{
```

File: src/Misc/Hooks.BugFixes.cpp [hunk 3]
```diff
@@ -636,6 +636,25 @@ DEFINE_HOOK(0x51A996, InfantryClass_PerCellProcess_KillOnImpassable, 0x5)
 	return SkipKilling;
 }
 
+// Fixes broken Upgrade logic to allow SpySat=Yes Code=Otamma / Tested DOOM3DGUY
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:639]: Code authorship should be put in credits, along with mentions in what's new and docs.

**Review by Metadorius — CHANGES_REQUESTED** [review]

**Comment by Starkku** [comment]
Calling it 'fixing broken upgrade logic' is bit misleading because there wasn't any handling for upgrade logic ever added beyond power drain/output, SuperWeapons and turrets/weapons, i.e what was used in Tiberian Sun.

File: docs/Whats-New.md [hunk 4]
```diff
@@ -306,6 +306,8 @@ New:
 - LaunchSW.DisplayMoney (by Starkku)
 - Disguise logic improvements (by Starkku)
 - Custom insignias (by Starkku)
+- Upgrade logic to allow altering of SpySat status (by Otamma)
+
```
Comment (Metadorius) [inline docs/Whats-New.md:310]:
```suggestion
- Upgrade logic to allow altering of SpySat status (by Otamaa)
```

File: CREDITS.md [hunk 5]
```diff
@@ -103,6 +103,7 @@ This page lists all the individual contributions to the project by their author.
   - Laser trails for VoxelAnims
   - MakeInfantry logic on BombClass bugfix
   - Debris & meteor impact behaviour settings
+  - Fixed broken Building Upgrade logic for SpySat=yes
```
Comment (Metadorius) [inline CREDITS.md:106]:
```suggestion
  - Upgrade logic to allow altering of SpySat status
```

**Review by Metadorius — APPROVED** [review]

**Pull request merged**